File: challenges/_templates/api/app/siem.py
"""Emision de eventos al SIEM collector (esquema ARCHITECTURE.md §5).

Fire-and-forget: el collector caido nunca rompe la request del jugador.
Copiable a cualquier reto Python.
"""
import json
import logging
import os
import threading
import urllib.request
from datetime import datetime, timezone

logger = logging.getLogger("siem")

COLLECTOR_URL = os.environ.get("COLLECTOR_URL", "http://collector:9000")
TEAM_ID = os.environ.get("TEAM_ID", "team_local")
CHALLENGE_ID = os.environ.get("CHALLENGE_ID", "unknown")


def _post(event: dict) -> None:
    try:
        data = json.dumps(event).encode()
        req = urllib.request.Request(
            f"{COLLECTOR_URL}/event", data=data,
            headers={"Content-Type": "application/json"}, method="POST",
        )
        urllib.request.urlopen(req, timeout=2.0)
    except Exception as exc:  # noqa: BLE001
        logger.warning("No se pudo enviar evento SIEM: %s", exc)


def emit(event_type: str, severity: str, src_ip: str | None = None, detail: dict | None = None) -> None:
    event = {
        "ts": datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ"),
        "source": "challenge",
        "team_id": TEAM_ID,
        "user": TEAM_ID,
        "src_ip": src_ip,
        "event_type": event_type,
        "severity": severity,
        "challenge_id": CHALLENGE_ID,
        "detail": detail or {},
    }
    threading.Thread(target=_post, args=(event,), daemon=True).start()

File: challenges/pwn/pwn-supply-04/forge/app/siem.js
/*
 * siem.js — emision de eventos SIEM (esquema ARCHITECTURE.md §5). Fire-and-forget.
 * Equivalente Node de siem.py. Sin dependencias externas.
 */
"use strict";
const http = require("http");
const { URL } = require("url");

const COLLECTOR_URL = process.env.COLLECTOR_URL || "http://collector:9000";
const TEAM_ID = process.env.TEAM_ID || "team_local";
const CHALLENGE_ID = process.env.CHALLENGE_ID || "pwn-supply-04";

function emit(eventType, severity, srcIp = null, detail = {}) {
  const event = {
    ts: new Date().toISOString().replace(/\.\d{3}Z$/, "Z"),
    source: "challenge",
    team_id: TEAM_ID,
    user: TEAM_ID,
    src_ip: srcIp,
    event_type: eventType,
    severity,
    challenge_id: CHALLENGE_ID,
    detail: detail || {},
  };
  try {
    const u = new URL(COLLECTOR_URL + "/event");
    const data = Buffer.from(JSON.stringify(event));
    const req = http.request(
      {
        hostname: u.hostname,
        port: u.port || 80,
        path: u.pathname,
        method: "POST",
        headers: { "Content-Type": "application/json", "Content-Length": data.length },
        timeout: 2000,
      },
      (res) => res.resume()
    );
    req.on("error", () => {});
    req.on("timeout", () => req.destroy());
    req.write(data);
    req.end();
  } catch (_) {
    /* fire-and-forget */
  }
}

module.exports = { emit };
